geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/pubsub/AbstractSubCommandsIntegrationTest.java

```diff
@@ -0,0 +1,203 @@
+
+    Runnable runnable = () -> subscriber.psubscribe(mockSubscriber, "f*");
+    Thread patternSubscriberThread = new Thread(runnable);
+
+    patternSubscriberThread.start();
+    waitFor(() -> mockSubscriber.getSubscribedChannels() == 1);
+    ArrayList<String> result =
+        (ArrayList<String>) introspector.pubsubChannels("f*");
+
+    assertThat(result).isEmpty();
+
+    mockSubscriber.punsubscribe();
+  }
+
+
+  @Test
+  public void channels_shouldReturnListOfMatchingChannels_withActiveChannelSubscribers_whenCalledWithPattern() {
+
+    ArrayList<String> expectedChannels = new ArrayList<>();
+    expectedChannels.add("foo");
+
+    Runnable runnable =
+        () -> subscriber.subscribe(mockSubscriber, "foo", "bar");
+
+    Thread subscriberThread = new Thread(runnable);
+    subscriberThread.start();
+    waitFor(() -> mockSubscriber.getSubscribedChannels() == 2);
+    ArrayList<String> result =
+        (ArrayList<String>) introspector.pubsubChannels("fo*");
+
+    assertThat(result).containsExactlyInAnyOrderElementsOf(expectedChannels);
+  }
+
+  @Test
+  public void channels_should_returnEmptyArry_GivenPatternWithNoMatches() {
+    ArrayList<String> result = (ArrayList<String>) introspector
+        .pubsubChannels("fo*");
+
+    assertThat(result).isEmpty();
+  }
+
+  @Test
+  @SuppressWarnings("unchecked")
+  public void channels_shouldOnlyReturnChannelsWithActiveSubscribers() {
+    ArrayList<byte[]> expectedChannels = new ArrayList<>();
+    expectedChannels.add("bar".getBytes());
+    Runnable runnable = () -> subscriber.subscribe(mockSubscriber, "foo", "bar");
+    Thread subscriberThread = new Thread(runnable);
+
+    subscriberThread.start();
+    waitFor(() -> mockSubscriber.getSubscribedChannels() == 2);
+    mockSubscriber.unsubscribe("foo");
+    waitFor(() -> mockSubscriber.getSubscribedChannels() == 1);
+    ArrayList<byte[]> result =
+        (ArrayList<byte[]>) introspector
+            .sendCommand(Protocol.Command.PUBSUB, "channels");
+
+    assertThat(result).containsExactlyInAnyOrderElementsOf(expectedChannels);
+  }
+
+  @Test
+  @SuppressWarnings("unchecked")
+  public void channels_shouldNotReturnDuplicates_givenMultipleSubscribersToSameChannel_whenCalledWithoutPattern() {
+
+    Jedis subscriber2 = new Jedis("localhost", getPort(), JEDIS_TIMEOUT);
+
+    MockSubscriber mockSubscriber2 = new MockSubscriber();
+    ArrayList<byte[]> expectedChannels = new ArrayList<>();
+    expectedChannels.add("foo".getBytes());
+
+    Runnable runnable = () -> subscriber.subscribe(mockSubscriber, "foo");
+    Thread subscriber1Thread = new Thread(runnable);
+    subscriber1Thread.start();
+
+    Runnable runnable2 = () -> subscriber2.subscribe(mockSubscriber2, "foo");
+    Thread subscriber2Thread = new Thread(runnable2);
+    subscriber2Thread.start();
+
+    waitFor(() -> {
+      return (mockSubscriber.getSubscribedChannels() == 1)
+          && (mockSubscriber2.getSubscribedChannels() == 1);
+    });
```

Comment:
This could be replaced with an expression lambda:
```
waitFor(() -> (mockSubscriber.getSubscribedChannels() == 1)
        && (mockSubscriber2.getSubscribedChannels() == 1));